airbyte-cdk/python/airbyte_cdk/sources/utils/transform.py

```diff
@@ -0,0 +1,190 @@
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in all
+# copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+#
+from distutils.util import strtobool
+from enum import Flag, auto
+from typing import Any, Callable, Dict
+
+from airbyte_cdk.logger import AirbyteLogger
+from jsonschema import Draft7Validator, validators
+
+logger = AirbyteLogger()
+
+
+class TransformConfig(Flag):
+    """
+    Transformer class config. Configs can be combined using bitwise or operator e.g.
+        ```
+        TransformConfig.DefaultSchemaNormalization | TransformConfig.CustomSchemaNormalization
+        ```
+    """
+
+    NoTransform = auto()
+    DefaultSchemaNormalization = auto()
+    CustomSchemaNormalization = auto()
+    # TODO: implement field transformation with user defined object path
+    FieldTransformation = auto()
+
+
+class Transformer:
+    """
+    Class for transforming object before output.
+    """
+
+    _custom_normalizer: Callable[[Any, Dict[str, Any]], Any] = None
+
+    def __init__(self, config: TransformConfig):
+        """
+        Initialize Transformer instance.
+        :param config Transform config that would be applied to object
+        """
+        if TransformConfig.NoTransform in config and config != TransformConfig.NoTransform:
+            raise Exception("NoTransform option cannot be combined with other flags.")
+        self._config = config
+        all_validators = {
+            key: self.__normalize_and_validate(key, orig_validator)
+            for key, orig_validator in Draft7Validator.VALIDATORS.items()
+            # Do not validate field we do not transform for maximum performance.
+            if key in ["type", "array", "$ref", "properties", "items"]
+        }
+        self._normalizer = validators.create(meta_schema=Draft7Validator.META_SCHEMA, validators=all_validators)
+
+    def register(self, normalization_callback: Callable) -> Callable:
+        """
+        Register custom normalization callback.
+        :param normalization_callback function to be used for value
+        normalization. Should return normalized value.
+        :return Same callbeck, this is usefull for using register function as decorator.
+        """
+        if TransformConfig.CustomSchemaNormalization not in self._config:
```

Comment:
what is the point of the user having to say upfront, `CustomSchemaNormalization`, if we have this `register` method? Why don't we just accept any normalization they register? 

I think the name `register` is also a bit misleading because you can only register exactly one trasnform. It should be called `setCustomTransform` instead

Reply:
Idea was that all transformation should be explicitly mentioned in stream config. e.g. if you see
```python
 transform = TypeTransformer(DefaultSchemaNormalization)
```
you know that only default transformation  applying  and there is no custom transformation registered to default transformer instance. If you want to register your own transformer you should explicitly config it for class instance.

Reply:
coming from the point of view of a developer leveraging this class, it's not clear what the contract of the registered normalizer is. What is the input to the callable I provide, and what should be the output?